tests/FunctionCallTest.php

```diff
@@ -433,52 +433,63 @@ public static function Baz($mixed) : string {
                      */
                     $elements = explode(" ", $string, $limit);',
                 'assertions' => [
-                    '$elements' => 'list<string>',
+                    '$elements' => 'list{0?: string, 1?: string, 2?: string, ...<int<0, max>, string>}',
                 ],
             ],
             'explodeWithDynamicDelimiter' => [
                 'code' => '<?php
                     /**
-                     * @var string $delim
+                     * @var non-empty-string $delim
                      * @var string $string
                      */
                     $elements = explode($delim, $string);',
                 'assertions' => [
-                    '$elements' => 'false|non-empty-list<string>',
+                    '$elements' => 'non-empty-list<string>',
+                ],
+            ],
+            'explodeWithDynamicDelimiterAndSmallPositiveLimit' => [
+                'code' => '<?php
+                    /**
+                     * @var non-empty-string $delim
+                     * @var string $string
+                     */
+                    $elements = explode($delim, $string, 2);',
+                'assertions' => [
+                    '$elements' => 'list{0: string, 1?: string}',
                 ],
             ],
             'explodeWithDynamicDelimiterAndPositiveLimit' => [
                 'code' => '<?php
                     /**
-                     * @var string $delim
+                     * @var non-empty-string $delim
                      * @var string $string
                      */
                     $elements = explode($delim, $string, 5);',
                 'assertions' => [
-                    '$elements' => 'false|non-empty-list<string>',
+                    '$elements' => 'non-empty-list<string>',
                 ],
             ],
             'explodeWithDynamicDelimiterAndNegativeLimit' => [
                 'code' => '<?php
                     /**
-                     * @var string $delim
+                     * @var non-empty-string $delim
                      * @var string $string
                      */
                     $elements = explode($delim, $string, -5);',
                 'assertions' => [
-                    '$elements' => 'false|list<string>',
+                    '$elements' => 'list<never>',
                 ],
             ],
             'explodeWithDynamicDelimiterAndLimit' => [
                 'code' => '<?php
                     /**
-                     * @var string $delim
+                     * @var non-empty-string $delim
                      * @var string $string
                      * @var int $limit
                      */
                     $elements = explode($delim, $string, $limit);',
                 'assertions' => [
-                    '$elements' => 'false|list<string>',
+                    '$elements' => 'list{0?: string, 1?: string, 2?: string, ...<int<0, max>, string>}',
```

Comment:
I wonder if adding an explicit  `$limit is int` case after the `$limit is 3` would prevent this smashing of type and still convey the info that `0` is definitely defined?

Reply:
I can try adding such a case, although tomorrow 😁

Reply:
I checked this: doesn't seem to have any effect (which is kinda expected)

Reply:
why is this expected? I understand that when the type is mixed, Psalm just merge all types together, but when it's a standard int, it should take the correct type, no?

Reply:
If I add an explicit `$limit is int` at the end of the conditional, it is equivalent to having the default case.